hubblestack/extmods/modules/audit.py

```diff
@@ -1,450 +1,726 @@
+        audit_result['check_result'] = check_result
+        return audit_result
+
+    # Check presence of implementation checks
+    if 'checks' not in audit_impl:
+        raise AuditCheckValidationError('No checks are present')
+
+    # Execute module validation of params
+    validate_param_method = audit_impl['module'] + '.validate_params'
+    for audit_check in audit_impl['checks']:
+        __nova__[validate_param_method](audit_id, audit_check)
+
+    # validate succeeded, lets execute it and prepare result dictionary
+    audit_result['run_config']['checks'] = []
+    execute_method = audit_impl['module'] + '.execute'
+    filtered_log_method = audit_impl['module'] + '.get_filtered_params_to_log'
+    # calculate the check result based on type parameter.
+    # If type is 'and', all subchecks should pass for success.
+    # If type is 'or', any passed subcheck will result in success.
+    overall_result = type == 'and'
+    failure_reasons = []
+    for audit_check in audit_impl['checks']:
+        if result_list is not None:
+            module_result_local = __nova__[execute_method](audit_id, audit_check, result_list)
+        else:
+            module_result_local = __nova__[execute_method](audit_id, audit_check)
+        audit_result_local = {}
+        if module_result_local['result']:
+            audit_result_local['check_result'] = CHECK_STATUS['Success']
+        else:
+            audit_result_local['check_result'] = CHECK_STATUS['Failure']
+            audit_result_local['failure_reason'] = module_result_local['failure_reason']
+            failure_reasons.append(audit_result_local['failure_reason'])
+        module_logs = {}
+        if not verbose:
+            log.debug('Non verbose mode')
+            module_logs = __nova__[filtered_log_method](audit_id, audit_check)
+        else:
+            log.debug('verbose mode')
+            module_logs = audit_check
+
+        audit_result_local = {**audit_result_local, **module_logs}
+        # add this result
+        audit_result['run_config']['checks'].append(audit_result_local)
+        if type == 'and':
+            overall_result = overall_result and module_result_local['result']
+        else:
+            overall_result = overall_result or module_result_local['result']
+
+    # Update overall check result based on invert result
+    overall_result = overall_result != invert_result
+
+    if overall_result:
+        audit_result['check_result'] = CHECK_STATUS['Success']
+    else:
+        # If check result is failure, fetch failure reason. If it is not present in profile, combine all individual checks reasons.
+        audit_result['check_result'] = CHECK_STATUS['Failure']
+        if failure_reason:
+            audit_result['failure_reason'] = failure_reason
+        else:
+            if failure_reasons:
+                failure_reasons = set(failure_reasons)
+                audit_result['failure_reason'] = ', '.join(failure_reasons)
```

Comment:
Let's just submit a list object as the failure reasons. It can be flattened into a comma separated string in Splunk if needed.
```suggestion
            audit_result['failure_reason'] = [failure_reason]
        else:
            if failure_reasons:
                failure_reasons = set(failure_reasons)
                audit_result['failure_reason'] = list(failure_reasons)
```